src/wasm-delegations-fields.h

```diff
@@ -39,65 +44,95 @@
 
 // Emits code to handle a child pointer.
 #ifndef DELEGATE_FIELD_CHILD
-#define DELEGATE_FIELD_CHILD(id, name)
+#error please define DELEGATE_FIELD_CHILD(id, name)
 #endif
 
-// Emits code to handle an optional child pointer (if this is not defined, then
-// DELEGATE_FIELD_CHILD is called on it).
+// Emits code to handle an optional child pointer. If this is not defined, then
+// DELEGATE_FIELD_CHILD is called on it.
 #ifndef DELEGATE_FIELD_OPTIONAL_CHILD
 #define DELEGATE_FIELD_OPTIONAL_CHILD(id, name) DELEGATE_FIELD_CHILD(id, name)
 #endif
 
-// Emits code to handle a variable-sized vector of child pointers.
+// Emits code to handle a variable-sized vector of child pointers. If this is
+// not defined, and DELEGATE_FIELD is, then DELEGATE_FIELD_CHILD is called on
+// them.
 #ifndef DELEGATE_FIELD_CHILD_VECTOR
-#define DELEGATE_FIELD_CHILD_VECTOR(id, name)
+#ifdef DELEGATE_GET_FIELD
+#define DELEGATE_FIELD_CHILD_VECTOR(id, name)                                  \
+  for (Index i = 0; i < (DELEGATE_GET_FIELD(id, name)).size(); i++) {          \
+    DELEGATE_FIELD_CHILD(id, name[i]);                                         \
+  }
+#else
+#error please define DELEGATE_FIELD_CHILD_VECTOR(id, name)
+#endif
 #endif
 
 // Emits code to handle an integer value (bool, enum, Index, int32, or int64).
 #ifndef DELEGATE_FIELD_INT
-#define DELEGATE_FIELD_INT(id, name)
+#error please define DELEGATE_FIELD_INT(id, name)
 #endif
 
 // Emits code to handle a std::array of fixed size of integer values (like a
-// SIMD mask).
+// SIMD mask). If this is not defined, and DELEGATE_FIELD is, then
+// DELEGATE_FIELD_INT is called on them.
 #ifndef DELEGATE_FIELD_INT_ARRAY
-#define DELEGATE_FIELD_INT_ARRAY(id, name)
+#ifdef DELEGATE_GET_FIELD
+#define DELEGATE_FIELD_INT_ARRAY(id, name)                                     \
+  for (Index i = 0; i < (DELEGATE_GET_FIELD(id, name)).size(); i++) {          \
+    DELEGATE_FIELD_INT(id, name[i]);                                           \
+  }
+#else
+#error please define DELEGATE_FIELD_INT_ARRAY(id, name)
+#endif
 #endif
 
 // Emits code to handle a Literal.
 #ifndef DELEGATE_FIELD_LITERAL
-#define DELEGATE_FIELD_LITERAL(id, name)
+#error please define DELEGATE_FIELD_LITERAL(id, name)
 #endif
 
 // Emits code to handle a name (like a call target).
 #ifndef DELEGATE_FIELD_NAME
-#define DELEGATE_FIELD_NAME(id, name)
+#error please define DELEGATE_FIELD_NAME(id, name)
+#endif
+
+// Emits code to handle a scope name definition (like a block's name).
+#ifndef DELEGATE_FIELD_SCOPE_NAME_DEF
+#error please define DELEGATE_FIELD_SCOPE_NAME_DEF(id, name)
 #endif
 
-// Emits code to handle a scope name (like a br target).
-#ifndef DELEGATE_FIELD_SCOPE_NAME
-#define DELEGATE_FIELD_SCOPE_NAME(id, name)
+// Emits code to handle a scope name use (like a br's target).
+#ifndef DELEGATE_FIELD_SCOPE_NAME_USE
+#error please define DELEGATE_FIELD_SCOPE_NAME_USE(id, name)
```

Comment:
Would provide an intermediate `define` be helpful, for the cases when we don't need to distinguish scope defs and uses? For example something like
```cpp
#ifndef DELEGATE_FIELD_SCOPE_NAME_DEF
#ifdef DELEGATE_FIELD_SCOPE_NAME
#define DELEGATE_FIELD_SCOPE_NAME_DEF DELEGATE_FIELD_SCOPE_NAME
#else
#error please define DELEGATE_FIELD_SCOPE_NAME_DEF(id, name)
#endif

#ifndef DELEGATE_FIELD_SCOPE_NAME_USE
#ifdef DELEGATE_FIELD_SCOPE_NAME
#define DELEGATE_FIELD_SCOPE_NAME_USE DELEGATE_FIELD_SCOPE_NAME
#else
#error please define DELEGATE_FIELD_SCOPE_NAME_USE(id, name)
#endif
```

Maybe we can extend this one step further so that we can provide a single `define` for all primitive types so that we don't need to write the same line for each `define` in the user code, like `DELEGATE_FIELD_PRIMITIVE` or something, and [all these](https://github.com/WebAssembly/binaryen/blob/ffdaad6adb8fa45f3c9021228b8e880ca02ff9b9/src/ir/ExpressionManipulator.cpp#L77-L83) can fall back to it when not defined, like
```cpp
// In wasm-delegations-fields.h
#ifndef DELEGATE_FIELD_LITERAL
#ifdef DELEGATE_FIELD_PRIMITIVE
#define DELEGATE_FIELD_LITERAL DELEGATE_FIELD_PRIMITIVE
#else
#error please define DELEGATE_LITERAL(id, name)
#endif

#ifndef DELEGATE_FIELD_SCOPE_NAME_DEF
#ifdef DELEGATE_FIELD_SCOPE_NAME
#define DELEGATE_FIELD_SCOPE_NAME_DEF DELEGATE_FIELD_SCOPE_NAME
#elif defined(DELEGATE_FIELD_PRIMITIVE)
#define DELEGATE_FIELD_SCOPE_NAME_DEF DELEGATE_FIELD_PRIMITIVE
#else
#error please define DELEGATE_FIELD_SCOPE_NAME_DEF(id, name)
#endif

// ... similar for other primitive fields ...

// In ExpressionManipulator.cpp
#define DELEGATE_FIELD_PRIMITIVE(id, name) castCopy->name = castOriginal->name;
```
But maybe this has gone too far. 😅 I'm not too serious here, and I'm just bikeshedding.

Reply:
That might make sense. Right now I don't think it would help much, but I'll be adding more uses of this code, so we can see if it helps later, and add it then perhaps?

Reply:
I would err on the side of keeping the delegation as simple as possible, especially since the documentation for this mechanism is currently just "read the code" and macro definitions aren't very readable readable. If we had separate documentation, I would be more excited about usability improvements like this.

Reply:
@kripken If you don't think it's helpful now, please ignore my comment. I was half bikeshedding anyway ;)

@tlively In general I was actually preferring un-erroring version on this header file when some of `define`'s are missing. I thought that user code can only make use of necessary `define`s would be more convenient, and we've seen many such def files and its uses in LLVM as well, in which some user code only makes use of opcode fields, while others use others.

This PR mandates definition, which in turn enforces user code to define the same line for many defines, so this was one idea to make user code simpler. But as I said I was half bikeshedding and I'm not very opinionated on this either way. Just explaining my rationale on this idea, that I preferred the un-erroring version 😅

Reply:
I think we can solve for all three of safety, usability, and discoverability, but it will require documentation separate from the macro soup :)

Reply:
I think a non-erroring version might be better for some things, like where forgetting a case would be detected in an obvious way otherwise. It is shorter, so we may want to have both eventually. I think it'll become clearer as we add more things.

@tlively what kind of docs did you have in mind? Something more than the comments before the macros begin?

Reply:
I think it would be sufficient to pull the usage notes for each of the possible macros up to the top of the file, where I could read them without getting distracted by their implementations.

Reply:
Sorry for flip-flopping, but given that we haven't fully decided on which path to take (non-erroring version? more macros with a separate doc? or none?), maybe it wouldn't be a good idea to provide this intermediate `DELEGATE_FIELD_SCOPE_NAME` at this point.

Reply:
@tlively I agree it is better for the docs that way, moving.

@aheejin sounds good, we can consider adding `DELEGATE_FIELD_SCOPE_NAME` later.